lex_test.go

```diff
@@ -0,0 +1,332 @@
+			input: ".local $hostName = {$host} .local $h = {$host}",
+			expected: []item{
+				// .local $hostName = {$host}
+				mk(itemKeyword, ".local"),
+				mk(itemWhitespace, " "),
+				mk(itemVariable, "$hostName"),
+				mk(itemWhitespace, " "),
+				mk(itemOperator, "="),
+				mk(itemWhitespace, " "),
+				mk(itemExpressionOpen, "{"),
+				mk(itemVariable, "$host"),
+				mk(itemExpressionClose, "}"),
+				mk(itemWhitespace, " "),
+				// .local $h = {$host}
+				mk(itemKeyword, ".local"),
+				mk(itemWhitespace, " "),
+				mk(itemVariable, "$h"),
+				mk(itemWhitespace, " "),
+				mk(itemOperator, "="),
+				mk(itemWhitespace, " "),
+				mk(itemExpressionOpen, "{"),
+				mk(itemVariable, "$host"),
+				mk(itemExpressionClose, "}"),
+				tokenEOF,
+			},
+		},
+		{
+			name:  "input declaration",
+			input: ".input {$host} .input {$user}",
+			expected: []item{
+				// .input {$host}
+				mk(itemKeyword, ".input"),
+				mk(itemWhitespace, " "),
+				mk(itemExpressionOpen, "{"),
+				mk(itemVariable, "$host"),
+				mk(itemExpressionClose, "}"),
+				mk(itemWhitespace, " "),
+				// .input ${user}
+				mk(itemKeyword, ".input"),
+				mk(itemWhitespace, " "),
+				mk(itemExpressionOpen, "{"),
+				mk(itemVariable, "$user"),
+				mk(itemExpressionClose, "}"),
+				tokenEOF,
+			},
+		},
+		{
+			name:  "reserved declaration",
+			input: ".output",
+			expected: []item{
+				mk(itemKeyword, ".output"),
+				tokenEOF,
+			},
+		},
+		{
+			name:  "matcher",
+			input: ".match {$n} 0 {{no apples}} 1 {{{$n} apple}} * {{{$n} apples}}",
+			expected: []item{
+				mk(itemKeyword, ".match"),
+				mk(itemWhitespace, " "),
+				tokenExpressionOpen,
+				mk(itemVariable, "$n"),
+				tokenExpressionClose,
+				mk(itemWhitespace, " "),
+				// 0 {{no apples}}
+				mk(itemLiteral, "0"),
+				mk(itemWhitespace, " "),
+				mk(itemQuotedPatternOpen, "{{"),
+				mk(itemText, "no apples"),
+				mk(itemQuotedPatternClose, "}}"),
+				mk(itemWhitespace, " "),
+				// 1 {{{$n} apple}}
+				mk(itemLiteral, "1"),
+				mk(itemWhitespace, " "),
+				mk(itemQuotedPatternOpen, "{{"),
+				mk(itemExpressionOpen, "{"),
+				mk(itemVariable, "$n"),
+				mk(itemExpressionClose, "}"),
+				mk(itemText, " apple"),
+				mk(itemQuotedPatternClose, "}}"),
+				mk(itemWhitespace, " "),
+				// * {{{$n} apples}}
+				mk(itemLiteral, "*"),
+				mk(itemWhitespace, " "),
+				mk(itemQuotedPatternOpen, "{{"),
+				mk(itemExpressionOpen, "{"),
+				mk(itemVariable, "$n"),
+				mk(itemExpressionClose, "}"),
+				mk(itemText, " apples"),
+				mk(itemQuotedPatternClose, "}}"),
+
+				tokenEOF,
+			},
+		},
+	} {
+		test := test
+
+		t.Run(test.name, func(t *testing.T) {
+			t.Parallel()
+
+			l := lex(test.input)
+
+			// collect items
+
+			var items []item
+
+			for {
+				v := l.nextItem()
+
+				logNext(t, l, v)
+
+				items = append(items, v)
+				if v.typ == itemEOF || v.typ == itemError {
+					break
+				}
+			}
+
+			// assert
+
+			assert.Equal(t, test.expected, items)
```

Comment:
how about, instead of checking slice versus slice, we check every element individually
`expected[i] == items[i]`

and if it fails then we call `lognext()`

Reason: decrease test output clutter

Reply:
Can you please elaborate on the clutter?

Unless you specify verbose output for tests you don't see it. Only when it fails.

Reply:
Yeah I see, but I thought we were running tests with `verbose` flag as the default, which seems not to be the case here.

Running without verbose flags, indeed logs this only when test fails.